src/fs/file.rs

```diff
@@ -222,6 +316,89 @@ impl File {
         op.write().await
     }
 
+    /// Attempts to write an entire buffer into this file at the specified offset.
+    ///
+    /// This method will continuously call [`write_at`] until there is no more data
+    /// to be written or an error of non-[`ErrorKind::Interrupted`] kind is returned.
+    /// This method will not return until the entire buffer has been successfully
+    /// written or such an error occurs.
+    ///
+    /// If the buffer contains no data, this will never call [`write_at`].
+    ///
+    /// # Return
+    ///
+    /// The method returns the operation result and the same buffer value passed
+    /// in as an argument.
+    ///
+    /// # Errors
+    ///
+    /// This function will return the first error of
+    /// non-[`ErrorKind::Interrupted`] kind that [`write_at`] returns.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use tokio_uring::fs::File;
+    ///
+    /// fn main() -> Result<(), Box<dyn std::error::Error>> {
+    ///     tokio_uring::start(async {
+    ///         let file = File::create("foo.txt").await?;
+    ///
+    ///         // Writes some prefix of the byte string, not necessarily all of it.
+    ///         let (res, _) = file.write_all_at(&b"some bytes"[..], 0).await;
+    ///         res?;
+    ///
+    ///         println!("wrote all bytes");
+    ///
+    ///         // Close the file
+    ///         file.close().await?;
+    ///         Ok(())
+    ///     })
+    /// }
+    /// ```
+    ///
+    /// [`write_at`]: File::write_at
+    /// [`ErrorKind::Interrupted`]: std::io::ErrorKind::Interrupted
+    pub async fn write_all_at<T: IoBuf>(&self, mut buf: T, pos: u64) -> crate::BufResult<(), T> {
+        let buf_len = buf.bytes_init();
+
+        if pos.checked_add(buf_len as u64).is_none() {
+            return (
+                Err(io::Error::new(
+                    io::ErrorKind::InvalidInput,
+                    "buffer too large for file",
+                )),
+                buf,
+            );
+        }
+
+        let mut bytes_written = 0;
+        while bytes_written < buf_len {
+            let (res, slice) = self
+                .write_at(buf.slice(bytes_written..), pos + bytes_written as u64)
+                .await;
+            buf = slice.into_inner();
+            match res {
+                Ok(0) => {
+                    return (
+                        Err(io::Error::new(
+                            io::ErrorKind::WriteZero,
+                            "failed to write whole buffer",
+                        )),
+                        buf,
+                    )
+                }
+                Ok(n) => {
+                    bytes_written += n;
+                }
+                Err(ref e) if e.kind() == io::ErrorKind::Interrupted => {}
+                Err(e) => return (Err(e), buf),
```

Comment:
Drive-by review: wouldn't it be more appropriate to return success if `bytes_written > 0`? Returning an error means the caller loses track of how much data was written out. In some scenarios that might be a security liability.

Same line of reasoning applies to the `Ok(0)` case.

Reply:
It's same as `std::fs::File::write_all` implementation, so I don't think this is a problem.